text/0000-cargo-rustup-discovery.md

```diff
@@ -0,0 +1,457 @@
+For `git`, this would be risky because users often run `git` from the shell prompt.
+This means that just changing directory to another user's directory could expose them.
+This may also apply to Rust users.
+For example [Starship](https://starship.rs/) can execute `rustc` to get its version for the shell prompt.
+
+## Filesystem behavior risks
+
+This fix assumes that a user can't in any way create a file owned by another user
+(unless they have elevated permissions like root).
+I believe this is the case for all major operating systems and filesystems.
+
+One (major) known exception is if a network-mounted filesystem is configured to map all files to the local user, which is not uncommon.
+In this case, the user will need to trust all users who can write to the mounted filesystem.
+
+## Check the ownership of every directory while traversing upwards
+
+The Git developers considered checking the ownership of every directory while traversing upwards,
+but they decided not to do it for performance reasons.
+I don't see a particularly strong reason to check the ownership along the way.
+I'd also be concerned about the performance consequence, though I expect it is extremely small.
+
+## Safe directory issues with Windows, WSL, and mingw
+
+On Windows, Git users have experienced some confusion and problems with setting safe directories.
+This is because git-for-windows is a mingw application,
+and needs to straddle the difference between POSIX and Windows style paths,
+as well as network paths.
+
+There is some concern that Cargo users using WSL or git-for-windows bash will have difficulty or confusion on how to set paths for the safe directories variable.
+I do not think it should be too much of an issue for Cargo because Cargo never uses POSIX paths, and only uses Windows-style paths.
+If the error message makes the full path clear, then the user should be able to copy-paste it.
+
+This may be particularly difficult for a user running Windows `cargo.exe` from within WSL.
+Paths in this environment look like `\\wsl$\Ubuntu-20.04\home\eric\foo`.
+I don't know how common this use case is, as I would expect someone doing things in WSL would install the native rustup/cargo binaries instead of using the ones from Windows.
+
+Another risk is entering Windows paths in a TOML config might be confusing.
+Either the user needs to use a literal string, or use escapes in a normal string.
+Some users have reported confusion about TOML string escaping.
+
+## Filesystem case-sensitivity
+
+I am proposing at this time that the safe directory check should be case-sensitive in all environments.
+However, this is causing issues for git users (see <https://github.com/git-for-windows/git/issues/3801>).
+
+This can manifest on Windows very easily.
+If you have a directory `Foo`, and you type `cd foo`,
+then the operating system will report the current directory is `foo`.
+
+I am not aware of an easy way to detect if a filesystem is case-sensitive.
```

Comment:
A typical case-sensitivity check creates two files with names that only differ in case. If the second create fails, or listing the directory only returns one file, then the filesystem is case-insensitive.

But it's faster to get the metadata of an existing file, via a path with a different case:
1. List the directory containing the file `a`
2. If the directory contains two entries for the file `a` that only differ by case, the filesystem is case-sensitive
3. Get the metadata of the file `A`, which has a name that only differs by case
4. If the query is successful, the filesystem is case-insensitive
5. Otherwise, if it returns a file not found error, the filesystem is case-sensitive

Unfortunately there's no easy API for this, because Windows can change case-sensitivity by directory, macOS by process (rarely), and most OSes can change it by filesystem.